source/agora/node/Ledger.d

```diff
@@ -579,17 +656,77 @@ public class Ledger
             [],
         );
 
+        ulong[Point] pkey_to_output_ind;
         // pay the commons budget
         if (commons_fee > Amount(0))
+        {
             coinbase_tx.outputs ~= Output(commons_fee,
                 this.params.CommonsBudgetAddress);
+            pkey_to_output_ind[this.params.CommonsBudgetAddress.data] = pkey_to_output_ind.length;
+        }
 
         // pay the validator for the past blocks
         if (auto payouts = this.fee_man.getAccumulatedFees(next_height))
             foreach (pair; payouts.byKeyValue())
                 if (pair.value > Amount(0))
+                {
                     coinbase_tx.outputs ~= Output(pair.value, pair.key);
+                    pkey_to_output_ind[pair.key.data] = pkey_to_output_ind.length;
+                }
+
+        ////////
+        // calculating block reward
+        ///////
+        if (block_reward_calc.is_payout_time(height))
+        {
+            immutable period_end = Height(height - params.BlockRewardDelay);
+            immutable period_start = Height(period_end - params.BlockRewardGap + 1);
+            Amount remainder;
+
+            // calculate the penalty for missing singatures
+            auto pkey_sigcounts = getPKeysOfSignedNonMissingValidators(period_start, period_end);
+            double total_actual_sigcount = pkey_sigcounts.byValue().fold!((a, b) => a + b)(0);
```

Comment:
Shouldn't this be an integer ? Why double ? Also I think you can use `sum` here ;)